modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/continuous/CacheContinuousQueryOperationFromCallbackTest.java

```diff
@@ -445,20 +574,46 @@ public CacheTestRemoteFilterAsync(String cacheName) {
         /** {@inheritDoc} */
         @Override public boolean evaluate(CacheEntryEvent<? extends QueryTestKey, ? extends QueryTestValue> e)
             throws CacheEntryListenerException {
-            if (e.getKey().compareTo(new QueryTestKey(KEYS)) < 0) {
+            if (e.getKey().key() < 0) {
                 IgniteCache<QueryTestKey, QueryTestValue> cache = ignite.cache(cacheName);
 
-                if (ThreadLocalRandom.current().nextBoolean()) {
-                    Set<QueryTestKey> keys = new LinkedHashSet<>();
+                boolean committed = false;
+                Transaction tx = null;
+                boolean startTx = cache.getConfiguration(CacheConfiguration.class).getAtomicityMode() != ATOMIC;
 
-                    for (int key = KEYS; key < KEYS + KEYS_FROM_CALLBACK; key++)
-                        keys.add(new QueryTestKey(key));
+                Set<QueryTestKey> keys = new LinkedHashSet<>();
 
-                    cache.invokeAll(keys, new IncrementTestEntryProcessor());
-                }
-                else {
-                    for (int key = KEYS; key < KEYS + KEYS_FROM_CALLBACK; key++)
-                        cache.invoke(new QueryTestKey(key), new IncrementTestEntryProcessor());
+                int startKey = ThreadLocalRandom.current().nextInt(KEYS_FROM_CALLBACK_RANGE - KEYS_FROM_CALLBACK);
+
+                for (int key = startKey; key < startKey + KEYS_FROM_CALLBACK; key++)
+                    keys.add(new QueryTestKey(key));
+
+                while (!committed && !Thread.currentThread().isInterrupted()) {
+                    try {
+                        if (startTx)
+                            tx = ignite.transactions().txStart(PESSIMISTIC, REPEATABLE_READ);
+
+                        if (ThreadLocalRandom.current().nextBoolean())
+                            cache.invokeAll(keys, new IncrementTestEntryProcessor());
+                        else {
+                            for (QueryTestKey key : keys)
+                                cache.invoke(key, new IncrementTestEntryProcessor());
+                        }
+
+                        if (tx != null)
+                            tx.commit();
+
+                        committed = true;
+                    }
+                    catch (Exception ex) {
+                        assertTrue(ex.getMessage(), ex.getMessage() != null &&
```

Comment:
Why do we expect TransactionRollback here?
Let's check cause for TransactionSerializationException.